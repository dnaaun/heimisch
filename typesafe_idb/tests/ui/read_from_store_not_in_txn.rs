use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
struct InstallationId(u64);

#[derive(Serialize, Deserialize)]
struct RepositoryId(u64);

#[derive(Serialize, Deserialize)]
struct UserId(u64);

#[derive(macros::TypesafeIdb, Serialize, Deserialize)]
struct Repository {
    #[idb(id)]
    id: RepositoryId,

    name: String,

    #[idb(index)]
    installation_id: InstallationId,
}

#[derive(macros::TypesafeIdb, Serialize, Deserialize)]
struct User {
    #[idb(id)]
    id: UserId,

    name: String,

    #[idb(index)]
    repository_id: RepositoryId,
}

pub fn main() {
    async {
        let db = typesafe_idb::TypesafeDb::builder("just test".into())
            .with_store::<Repository>()
            .with_store::<User>()
            .build()
            .await
            .unwrap();

        let txn = db.txn().with_store::<User>().ro();

        // Should raise type error
        txn.object_store::<Repository>().unwrap();
    };
}
