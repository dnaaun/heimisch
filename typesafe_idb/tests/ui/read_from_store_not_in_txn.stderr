error[E0277]: the trait bound `typesafe_idb::chain::Chain<UserStoreMarker, typesafe_idb::chain::Chain<(), ()>>: StoreMarker<Repository>` is not satisfied
  --> tests/ui/read_from_store_not_in_txn.rs:46:13
   |
46 |         txn.object_store::<Repository>().unwrap();
   |             ^^^^^^^^^^^^ the trait `StoreMarker<Repository>` is not implemented for `typesafe_idb::chain::Chain<UserStoreMarker, typesafe_idb::chain::Chain<(), ()>>`
   |
   = help: the trait `StoreMarker<S>` is implemented for `typesafe_idb::chain::Chain<Head, Tail>`
note: required by a bound in `Txn::<Markers, Mode>::object_store`
  --> src/txn.rs
   |
   |     pub fn object_store<S>(&self) -> Result<ObjectStore<'_, S, Mode>, crate::Error>
   |            ------------ required by a bound in this associated function
...
   |         Markers: StoreMarker<S>,
   |                  ^^^^^^^^^^^^^^ required by this bound in `Txn::<Markers, Mode>::object_store`
